Make mfenced render like an mrow.

eerii opened the pull request:
#### 8fbb30da9f209cb85ad6809cf259b25dc3b2af85
<pre>
Make mfenced render like an mrow.
<a href="https://bugs.webkit.org/show_bug.cgi?id=160509">https://bugs.webkit.org/show_bug.cgi?id=160509</a>

Reviewed by Frédéric Wang.

Remove the previous special cases so that mfenced renders as an mrow.
Gated under the CoreMathMLEnabled feature.

As of this patch, fences are no longer exposed in the a11y trees for
mfenced when CoreMathMLEnabled is active.

Tests: mathml/non-core/fenced-mi.html
       mathml/non-core/fenced.html
* LayoutTests/TestExpectations:
* LayoutTests/accessibility/math-fenced.html:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/legacy-mfenced-element-001.html:
* LayoutTests/mathml/non-core/fenced-expected.html: Renamed from LayoutTests/mathml/presentation/fenced-expected.html.
* LayoutTests/mathml/non-core/fenced-mi-expected.html: Renamed from LayoutTests/mathml/presentation/fenced-mi-expected.html.
* LayoutTests/mathml/non-core/fenced-mi.html: Renamed from LayoutTests/mathml/presentation/fenced-mi.html.
* LayoutTests/mathml/non-core/fenced.html: Renamed from LayoutTests/mathml/presentation/fenced.html.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/operators/embellished-operator-001-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::addChildren):
* Source/WebCore/mathml/MathMLRowElement.cpp:
(WebCore::MathMLRowElement::createElementRenderer):

Canonical link: <a href="https://commits.webkit.org/300587@main">https://commits.webkit.org/300587@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/946cebda4e6d0a6300444d32825060b4e38e251f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123101 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42815 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129760 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75211 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124978 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43539 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51410 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93571 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62092 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126052 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34699 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110172 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74207 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33667 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28328 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73275 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104411 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28554 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132489 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50051 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38101 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102072 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106393 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101925 "Found 2 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25917 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47298 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25496 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46824 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49906 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55667 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49374 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52726 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51055 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented:
@eerii Can we do that under a runtime pref?

cc @Ahmad-S792

eerii commented:
> Can we do that under a runtime pref?

Yes, definitely

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c2364d7677b33c16c1e51b25dd510b59b21361d9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51100

fred-wang commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -1834,6 +1834,11 @@ imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2.html [ Sk
+# These internal MathML tests are outdated and do not match MathML Core.
+accessibility/math-fenced.html [ Failure ]
+mathml/presentation/fenced.html [ ImageOnlyFailure ]
+mathml/presentation/fenced-mi.html [ ImageOnlyFailure ]
```
Can you move these two to `LayoutTests/mathml/non-core/` and add a  `<!-- webkit-test-runner [ CoreMathMLEnabled=false ] -->` flag to them?

fred-wang commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -1834,6 +1834,11 @@ imported/w3c/web-platform-tests/mathml/relations/css-styling/lengths-2.html [ Sk
 webkit.org/b/212080 mathml/presentation/href-style.html [ ImageOnlyFailure ]
 
+# These internal MathML tests are outdated and do not match MathML Core.
+accessibility/math-fenced.html [ Failure ]
```
Can we keep this test passing if we preverve the a11y behavior (see below)?

fred-wang commented on Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
```diff
@@ -71,7 +71,7 @@ AccessibilityRole AccessibilityMathMLElement::determineAccessibilityRole()
 void AccessibilityMathMLElement::addChildren()
 {
-    if (!hasElementName(ElementName::MathML_mfenced)) {
+    if ((document() && document()->settings().coreMathMLEnabled()) || !hasElementName(ElementName::MathML_mfenced)) {
```
I think the way we expose mfenced to AT is covered by https://searchfox.org/wubkat/search?q=math-fenced&path=&case=true&regexp=false

If you don't do this change, would this test still passes?

eerii replied:
I checked Firefox and Chromium and neither of them exposes fences either. I guess this makes some sense, otherwise users using the AT would see something different than what's being displayed, and with MathML Core mfenced is just an unknown element.

I added a note in the commit description and I agree with running the test like we did before.

<img height="150" alt="Chromium accessibility tree for mfenced" src="https://github.com/user-attachments/assets/0c430e51-2d46-42aa-a15a-1f2947457621" />

<img height="150" alt="Firefox accessibility tree for mfenced" src="https://github.com/user-attachments/assets/c821eb45-3e2e-45c5-87f7-47a3ba2ee7c6" />

fred-wang commented on Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
```diff
@@ -365,9 +365,11 @@ void RenderTreeBuilder::attachInternal(RenderElement& parent, RenderPtr<RenderOb
-    if (auto* mathMLFenced = dynamicDowncast<RenderMathMLFenced>(parent)) {
-        mathMLBuilder().attach(*mathMLFenced, WTFMove(child), beforeChild);
-        return;
+    if (!child || !child->document().settings().coreMathMLEnabled()) {
```
why are you testing testing `!child?`

Also, I think we are no longer creating RenderMathMLFenced when the flag is enabled, so is this change necessary?

eerii replied:
Yes, this is no longer necessary at all. I'll remove it. Thanks

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6f610bc13966d1ca9ca59680e0a18cea769ad7df. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51100

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/25e5dda1f140cece0c91d61c63f59fff56f99530. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51100

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b66ffa8012769842c289c3a920ea94af0aa4d143. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51100

fred-wang commented on LayoutTests/platform/glib/accessibility/math-fenced-expected.txt:
```diff
@@ -5,13 +5,9 @@ On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE
         AXRole: AXStatic AXValue: a
-        AXRole: AXStatic AXValue: ;
         AXRole: AXStatic AXValue: e
-        AXRole: AXStatic AXValue: }
```
Can you please instead add `<!-- webkit-test-runner [ CoreMathMLEnabled=false ] -->` to `LayoutTests/accessibility/math-fenced.html` too? The goal is really to check that the open/close/separators attributes are exposed to AT so we should really run it with the flag enabled...

Maybe add a comment in `math-fenced.html` that this is testing support for `mfenced`, which is deprecated and not part of MathML Core?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/df69fde0f03f56b15bf1e06bfdf35966c0bdbdfa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51100

eerii committed:
Make mfenced render like an mrow.
https://bugs.webkit.org/show_bug.cgi?id=160509

Reviewed by Frédéric Wang.

Remove the previous special cases so that mfenced renders as an mrow.
Gated under the CoreMathMLEnabled feature.

As of this patch, fences are no longer exposed in the a11y trees for
mfenced when CoreMathMLEnabled is active.

Tests: mathml/non-core/fenced-mi.html
       mathml/non-core/fenced.html
* LayoutTests/TestExpectations:
* LayoutTests/accessibility/math-fenced.html:
* LayoutTests/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/legacy-mfenced-element-001.html:
* LayoutTests/mathml/non-core/fenced-expected.html: Renamed from LayoutTests/mathml/presentation/fenced-expected.html.
* LayoutTests/mathml/non-core/fenced-mi-expected.html: Renamed from LayoutTests/mathml/presentation/fenced-mi-expected.html.
* LayoutTests/mathml/non-core/fenced-mi.html: Renamed from LayoutTests/mathml/presentation/fenced-mi.html.
* LayoutTests/mathml/non-core/fenced.html: Renamed from LayoutTests/mathml/presentation/fenced.html.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/mrow/mrow-preferred-width-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/presentation-markup/operators/embellished-operator-001-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::addChildren):
* Source/WebCore/mathml/MathMLRowElement.cpp:
(WebCore::MathMLRowElement::createElementRenderer):

Canonical link: https://commits.webkit.org/300587@main
